operator/pkg/clustermesh/clustermesh.go

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package clustermesh
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/cilium/endpointslice-controller/endpointslice"
+	"k8s.io/client-go/informers"
+	cache "k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/clustermesh"
+	"github.com/cilium/cilium/pkg/clustermesh/common"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/kvstore/store"
+	serviceStore "github.com/cilium/cilium/pkg/service/store"
+)
+
+// clusterMesh is a cache of multiple remote clusters
+type clusterMesh struct {
+	// common implements the common logic to connect to remote clusters.
+	common common.ClusterMesh
+
+	context       context.Context
+	contextCancel context.CancelFunc
+	Metrics       Metrics
+
+	// globalServices is a list of all global services. The datastructure
+	// is protected by its own mutex inside the structure.
+	globalServices *globalServiceCache
+
+	storeFactory store.Factory
+
+	conurrentClusterMeshEndpointSync int
+
+	meshPodInformer     *meshPodInformer
+	meshNodeInformer    *meshNodeInformer
+	meshServiceInformer *meshServiceInformer
+
+	endpointSliceMeshController  *endpointslice.Controller
+	endpointSliceInformerFactory informers.SharedInformerFactory
+}
+
+func newClusterMesh(lc cell.Lifecycle, params clusterMeshParams) *clusterMesh {
+	if !params.Clientset.IsEnabled() || params.ClusterMeshConfig == "" {
+		return nil
+	}
+
+	log.WithField("enabled", params.Cfg.ClusterMeshEnableEndpointSync).
+		Info("Endpoint Slice Cluster Mesh synchronization")
+
+	if !params.Cfg.ClusterMeshEnableEndpointSync {
+		return nil
+	}
+
+	cm := clusterMesh{
+		Metrics: params.Metrics,
+		globalServices: newGlobalServiceCache(
+			params.Metrics.TotalGlobalServices.WithLabelValues(params.ClusterInfo.Name),
+		),
+		storeFactory:                     params.StoreFactory,
+		conurrentClusterMeshEndpointSync: params.Cfg.ClusterMeshMaxEndpointsPerSlice,
+	}
+	cm.context, cm.contextCancel = context.WithCancel(context.Background())
+	cm.meshPodInformer = &meshPodInformer{GlobalServiceCache: cm.globalServices}
+	cm.meshNodeInformer = newMeshNodeInformer()
+	cm.endpointSliceMeshController, cm.meshServiceInformer, cm.endpointSliceInformerFactory = newEndpointSliceMeshController(
+		cm.context, params.Cfg, cm.meshPodInformer,
+		cm.meshNodeInformer, params.Clientset,
+		params.Services, cm.globalServices,
+	)
+	cm.common = common.NewClusterMesh(common.Configuration{
```

Comment:
Ah! I assumed the operator did not ran in HostNetwork :sweat_smile:

Reply:
It does, as it needs to start before the agents. While the clustermesh-apiserver does run in pod network.

Reply:
Ohh ok I see, TIL, thanks for the explanation :D! I will look into the `ServiceIpGetter` then :+1:

Reply:
Hmmm IIUC in the clustermesh package this is provided by the ServiceCache Cell which is not available into the operator it seems, should I add it or is not appropriate for the operator? Sorry I am not super familiar with this part of the code :/

Reply:
Hmm, that part is pretty messy. The easiest solution seems to implement the `GetServiceIP` interface on top of the `resource.Resource[*slim_corev1.Service]` store. Just lookup the service and return the corresponding IP/port (I'm actually confused about why the port is also retrieved from the service, instead of just being taken from the original url, but that [seems the case today](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/pkg/k8s/service.go#L711-L716)).

Reply:
I added a wip commit to do that let me know if it what you had in mind!

Reply:
Yep, looks reasonable to me. Thanks!